hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -317,31 +362,48 @@ protected List<DatanodeDetails> chooseDatanodesInternal(
                       "available racks. {} Available racks count: {},"
                       + " Excluded nodes count: {}, UsedNodes count: {}",
               reason, racks.size(), excludedNodesCount, usedNodesCount);
-      throw new SCMException(reason,
-              SCMException.ResultCodes.FAILED_TO_FIND_HEALTHY_NODES);
+
+      /*
+      Couldn't choose enough racks to satisfy the rack scatter policy. If
+      fallback is not allowed, then throw an exception.
+      */
+      if (!fallback) {
+        throw new SCMException(reason,
+            SCMException.ResultCodes.FAILED_TO_FIND_HEALTHY_NODES);
+      }
     }
 
     if (chosenNodes.size() < nodesRequired) {
+      /* If fallback is allowed, adjust maxReplicasToMove such that
+      replication factor number of replicas are allowed per rack.
+      */
+      if (fallback) {
+        maxReplicasPerRack = requiredReplicationFactor;
+      }
       racks.addAll(usedRacksCntMap.keySet());
       racks = sortRackWithExcludedNodes(racks, excludedNodes, usedRacksCntMap);
       racks.addAll(usedRacksCntMap.keySet());
+      LOG.debug("Available racks considering racks with used and exclude " +
+              "nodes: {}.", racks);
       chosenNodes.addAll(chooseNodesFromRacks(racks, unavailableNodes,
               mutableFavoredNodes, nodesRequired - chosenNodes.size(),
               metadataSizeRequired, dataSizeRequired,
-              Integer.MAX_VALUE, usedRacksCntMap, maxReplicasPerRack));
+              nodesRequired - chosenNodes.size(),
```

Comment:
Good point. We can set `maxOuterLoopIterations` to a higher value. Looks like the number of iterations are bounded by both `maxOuterLoopIterations` and `OUTER_LOOP_MAX_RETRY`. The latter is a constant and its value currently is `3`, so we need to increase that too. Any suggestions for what their values should be?

Reply:
OUTER_LOOP_MAX_RETRY is there, when we don't find any nodes in any rack even after going through all the racks OUTER_LOOP_MAX_RETRY times, I believe we can make it configurable but default value seems to be fine. Even if we find one node in the iteration retry count is reset.

Reply:
Keeping `INT_MAX` doesn't harm anything here as `OUTER_LOOP_MAX_RETRY` already controls for exit condition too in case we really could not find enough nodes.

Reply:
Ok, the consensus is that we should keep `maxOuterLoopIterations` as `Integer.MAX_VALUE`. I wonder if we should then change this argument to `Integer.MAX_VALUE` when we call `chooseNodesFromRacks` at line 352, too?

Reply:
At line 352 it should be max replica per rack, If we increase this iteration count beyond max replica per rack it can use same rack for multiple replica because every time outer loop will start reusing all the racks. Only in the same iteration it removes rack from reusing.
If we want to optimize, we can remove racks which has already reached max replica and then any number of times like INT_MAX we can do for outer loop.

Reply:
> If we want to optimize, we can remove racks which has already reached max replica

I agree, and I think we're already doing that.
At line 70, we update the count of used nodes per rack:
```
rackCntMap.merge(rack, 1, Math::addExact);
```
And at line 126, we filter nodes according to this map:
```
      for (Node rack : racks) {
        if (rackCntMap.getOrDefault(rack, 0) < maxReplicasPerRack) {
          toChooseRacks.add(rack);
        }
      }

```

Reply:
I've decided to make it `maxReplicasPerRack` at line 352 for the first call, and `Integer.MAX_VALUE ` for the second call. If we want to increase it for the first call, we can do that in this separate jira - https://issues.apache.org/jira/browse/HDDS-9054.

Reply:
Looks fine now, earlier it was 1 for the 1st call now changed to `maxReplicasPerRack` which is better than 1 and don't have any impact. Yes later we can optimize more in new Jira you created HDDS-9054.
You can update PR description as well for `maxOuterLoopIterations`.

Reply:
Thanks @ashishkumar50 for your review. I've now updated the PR description as well.